pkg/mapper/metadata.go

```diff
@@ -0,0 +1,178 @@
+// If there are no groups, it returns an empty slice.
+// Usage example: systemMetadata := datum.SystemMetadata()
+// groups := systemMetadata.Groups()
+func (md SystemMetadata) Groups() []string {
+	groups := make([]string, 0, len(md.data))
+	for group := range md.data {
+		groups = append(groups, group)
+	}
+	return groups
+}
+
+// Keys returns the keys of the system metadata for the given group.
+// If the group is not present, it returns an empty slice.
+// Usage example: systemMetadata := datum.SystemMetadata()
+// keys := systemMetadata.Keys("group-name")
+func (md SystemMetadata) Keys(group string) []string {
+	keys := make([]string, 0, len(md.data[group]))
+	for key := range md.data[group] {
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+// Value returns the value of the system metadata for the given group and key.
+// If the group or key is not present, it returns an empty slice.
+// Usage example: systemMetadata := datum.SystemMetadata()
+// value := systemMetadata.Value("group-name", "key")
+func (md SystemMetadata) Value(group, key string) []byte {
+	return md.data[group][key]
+}
+
+// UserMetadata wraps user-defined metadata groups per message.
+type UserMetadata struct {
+	data map[string]map[string][]byte
+}
+
+// NewUserMetadata wraps an existing map into UserMetadata.
+// If d is nil, an empty map is created.
+func NewUserMetadata(d map[string]map[string][]byte) UserMetadata {
+	if d == nil {
+		d = make(map[string]map[string][]byte)
+	}
+	return UserMetadata{data: d}
+}
+
+// Groups returns the groups of the user metadata.
+// If there are no groups, it returns an empty slice.
+// Usage example: userMetadata := datum.UserMetadata()
+// groups := userMetadata.Groups()
+func (md UserMetadata) Groups() []string {
+	groups := make([]string, 0, len(md.data))
+	for group := range md.data {
+		groups = append(groups, group)
+	}
+	return groups
+}
+
+// Keys returns the keys of the user metadata for the given group.
+// If the group is not present, it returns an empty slice.
+// Usage example: userMetadata := datum.UserMetadata()
+// keys := userMetadata.Keys("group-name")
+func (md UserMetadata) Keys(group string) []string {
+	keys := make([]string, 0, len(md.data[group]))
+	for key := range md.data[group] {
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+// Value returns the value of the user metadata for the given group and key.
+// If the group or key is not present, it returns an empty slice.
+// Usage example: userMetadata := datum.UserMetadata()
+// value := userMetadata.Value("group-name", "key")
+func (md UserMetadata) Value(group, key string) []byte {
+	return md.data[group][key]
+}
+
+// SetKVGroup sets a group of key-value pairs under the provided group name.
+// Usage example: userMetadata := NewUserMetadata()
+// userMetadata.SetKVGroup("group-name", map[string][]byte{"key": []byte("value")})
+func (md UserMetadata) SetKVGroup(group string, kv map[string][]byte) {
```

Comment:
It's better to use pointer receiver for all methods https://dave.cheney.net/2016/03/19/should-methods-be-declared-on-t-or-t
Value receivers can easily create bugs. Eg
```go
type Metadata struct {
	m map[string]string
}

func(m Metadata) Set(k, v string) {
	if m.m == nil {
		m.m = make(map[string]string)
	}
	m.m[k] = v
}

func main() {
	var m Metadata
	m.Set("a", "b") // mutates the copy
	fmt.Println(m.m) // map[]
}
```

If we later have a need to introduce a private field  (basic data types like `int`) in `UserMetadata` and mutate it in public methods, it won't be possible. Also, some other data types like `sync.Mutex` or `sync.WaitGroup` should not be copied. Value receiver will restrict us from using these.